### DISCUSSION TITLE
Faster approximation to Steiner tree problem

### DISCUSSION THREAD

**Discussion by GuyAglionby** [networkx/networkx#5352] [Algorithms]
Background: the Steiner tree problem given a graph G = (V, E) and a subset of its nodes S is to find the minimal set of edges in G that span S.

NetworkX has an [approximation algorithm for this](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.approximation.steinertree.steiner_tree.html#networkx.algorithms.approximation.steinertree.steiner_tree) with a runtime of O(|V|^2 |S|), I think the one from [Kou et al.](https://link.springer.com/article/10.1007/BF00288961).

As V grows, this quickly becomes intractable (I have a graph with 2 million nodes..!). Fortunately, there are other, faster approximation algorithms for the problem that have the same error bounds. In particular, I implemented and am using the algorithm from [Wu et al.](https://link.springer.com/article/10.1007/BF00289500) that runs in O(|E| log |V|).

I'm looking to contribute this, but had a couple questions. I'm not sure whether this should replace the existing algorithm, or be offered as an alternative? If the latter I'm not sure what the naming conventions are. Happy to open a PR, let me know, cheers!

**Comment by boothby** [comment]
This sounds great, and I'd be happy to review that PR.  Note that the generic shortest_path function's implementation is controlled with the `method` parameter, and I believe that to be standard.

https://github.com/networkx/networkx/blob/42985ba7d9f768c32c651e3e73d4d98b46776f54/networkx/algorithms/shortest_paths/generic.py#L39

Since the two algorithms achieve a similar approximation ratio, I'm weakly inclined to toss out the old in favor of the new.  But, I have two reasons to maintain both methods.  First, I'm unsure of the quality differential on small graphs, and there might be a tangible benefit to taking the minimum result of running both algorithms when time is not an issue.  Second, NetworkX is in part an educational project; students of graph theory benefit from having clear implementations of important algorithms here.

**Reply by GuyAglionby** [reply]
Great! Thanks for the answers here. I'll send over a PR in due course
